internal/radix/adaptive/helpers.go

```diff
@@ -0,0 +1,737 @@
+			leaf, ok := n.(*NodeLeaf[T])
+			if !ok {
+				continue
+			}
+			// Check if the expanded path matches
+			if leafMatches[T](leaf, key, keyLen) == 0 {
+				return leaf.value, true
+			}
+			return zero, false
+		}
+
+		// Bail if the prefix does not match
+		if n.getPartialLen() > 0 {
+			prefixLen := checkPrefix[T](n, key, keyLen, depth)
+			if prefixLen != min(MaxPrefixLen, int(n.getPartialLen())) {
+				return zero, false
+			}
+			depth += int(n.getPartialLen())
+		}
+
+		if depth >= keyLen {
+			return zero, false
+		}
+
+		// Recursively search
+		child = t.findChild(n, key[depth])
+		if child != nil && *child != nil && **child != nil {
+			n = **child
+		} else {
+			n = nil
+		}
+		depth++
+	}
+	return zero, false
+}
+
+func checkPrefix[T any](n Node[T], key []byte, keyLen, depth int) int {
+	maxCmp := min(min(int(n.getPartialLen()), MaxPrefixLen), keyLen-depth)
+	var idx int
+	for idx = 0; idx < maxCmp; idx++ {
+		if n.getPartial()[idx] != key[depth+idx] {
+			return idx
+		}
+	}
+	return idx
+}
+
+func leafMatches[T any](n *NodeLeaf[T], key []byte, keyLen int) int {
+	// Fail if the key lengths are different
+	if int(n.keyLen) != keyLen {
+		return 1
+	}
+	// Compare the keys
+	return bytes.Compare(n.key, key)
+}
+
+func (t *RadixTree[T]) recursiveInsert(n *Node[T], ref **Node[T], key []byte, value T, depth int, old *int) T {
+	var zero T
+	keyLen := len(key)
+
+	// If we are at a nil node, inject a leaf
+	if n == nil {
+		leafNode := t.makeLeaf(key, value)
+		*ref = &leafNode
+		return zero
+	}
+
+	node := *n
+	if node.isLeaf() {
+		nodeLeaf := node.(*NodeLeaf[T])
+		// This means root is nil
+		if len(nodeLeaf.key) == 0 {
+			leafNode := t.makeLeaf(key, value)
+			*ref = &leafNode
+			return zero
+		}
+	}
+
+	// If we are at a leaf, we need to replace it with a node
+	if node.isLeaf() {
+		nodeLeaf := node.(*NodeLeaf[T])
+
+		// Check if we are updating an existing value
+		if bytes.Equal(nodeLeaf.key, key) {
+			*old = 1
+			oldVal := nodeLeaf.value
+			nodeLeaf.value = value
+			return oldVal
+		}
+
+		// New value, we must split the leaf into a node4
+		newLeaf2 := t.makeLeaf(key, value).(*NodeLeaf[T])
+
+		// Determine longest prefix
+		longestPrefix := longestCommonPrefix[T](nodeLeaf, newLeaf2, depth)
+		newNode := t.allocNode(NODE4)
+		newNode4 := newNode.(*Node4[T])
+		newNode4.partialLen = uint32(longestPrefix)
+		copy(newNode4.partial[:], key[depth:depth+min(MaxPrefixLen, longestPrefix)])
+
+		// Add the leafs to the new node4
+		t.addChild4(newNode4, ref, nodeLeaf.key[depth+longestPrefix], nodeLeaf)
+		t.addChild4(newNode4, ref, newLeaf2.key[depth+longestPrefix], newLeaf2)
+		*ref = &newNode
+		return zero
+	}
+
+	// Check if given node has a prefix
+	if node.getPartialLen() > 0 {
+		// Determine if the prefixes differ, since we need to split
+		prefixDiff := prefixMismatch[T](node, key, keyLen, depth)
+		if prefixDiff >= int(node.getPartialLen()) {
+			depth += int(node.getPartialLen())
+			goto RECURSE_SEARCH
```

Comment:
Gotos make reading and understanding code harder and should be avoided if possible. 
In this case it looks like its being used to avoid indenting lines 140-166 in an `else` block. So you could put all that into an `else` block instead, but it likely would be better to extract some other functions out of this one and call them instead of inlining all their code. That should help to improve readability of the code and make it clearer what the overall control flow looks like from a higher level.